// gonzales-pe/lib/less/node-types.js
module.exports = {
    ArgumentsType: 'arguments',
    AtkeywordType: 'atkeyword',
    AtrulebType: 'atruleb',
    AtrulerType: 'atruler',
    AtrulerqType: 'atrulerq',
    AtrulersType: 'atrulers',
    AtrulesType: 'atrules',
    AttribType: 'attrib',
    AttrselectorType: 'attrselector',
    BlockType: 'block',
    BracesType: 'braces',
    ClassType: 'class',
    CombinatorType: 'combinator',
    CommentMLType: 'commentML',
    CommentSLType: 'commentSL',
    ConditionType: 'condition',
    DeclarationType: 'declaration',
    DeclDelimType: 'declDelim',
    DelimType: 'delim',
    DimensionType: 'dimension',
    EscapedStringType: 'escapedString',
    FilterType: 'filter',
    FiltervType: 'filterv',
    FunctionType: 'function',
    FunctionExpressionType: 'functionExpression',
    IdentType: 'ident',
    ImportantType: 'important',
    IncludeType :'include',
    InterpolatedVariableType: 'interpolatedVariable',
    MixinType: 'mixin',
    NamespaceType: 'namespace',
    NthType: 'nth',
    NthselectorType: 'nthselector',
    NumberType: 'number',
    OperatorType: 'operator',
    ParentSelectorType: 'parentselector',
    PercentageType: 'percentage',
    ProgidType: 'progid',
    PropertyType: 'property',
    PropertyDelimType: 'propertyDelim',
    PseudocType: 'pseudoc',
    PseudoeType: 'pseudoe',
    RawType: 'raw',
    RulesetType: 'ruleset',
    SType: 's',
    SelectorType: 'selector',
    ShashType: 'shash',
    SimpleselectorType: 'simpleselector',
    StringType: 'string',
    StylesheetType: 'stylesheet',
    UnaryType: 'unary',
    UriType: 'uri',
    ValueType: 'value',
    VariableType: 'variable',
    VariablesListType: 'variableslist',
    VhashType: 'vhash'
};
